plugins/inputs/elasticsearch_query/elasticsearch_query.go

```diff
@@ -0,0 +1,277 @@
+package elasticsearch_query
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"net/http"
+	"os"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+
+	elastic "gopkg.in/olivere/elastic.v5"
+
+	"github.com/influxdata/telegraf"
+	"github.com/influxdata/telegraf/internal"
+	"github.com/influxdata/telegraf/internal/tls"
+	"github.com/influxdata/telegraf/plugins/inputs"
+)
+
+const description = `Queries Elasticseach`
+const sampleConfig = `
+  ## The full HTTP endpoint URL for your Elasticsearch instance
+  ## Multiple urls can be specified as part of the same cluster,
+  ## this means that only ONE of the urls will be written to each interval.
+  urls = [ "http://node1.es.example.com:9200" ] # required.
+  ## Elasticsearch client timeout, defaults to "5s" if not set.
+  timeout = "5s"
+  ## Set to true to ask Elasticsearch a list of all cluster nodes,
+  ## thus it is not necessary to list all nodes in the urls config option
+  enable_sniffer = false
+  ## Set the interval to check if the Elasticsearch nodes are available
+  ## Setting to "0s" will disable the health check (not recommended in production)
+  health_check_interval = "10s"
+  ## HTTP basic authentication details (eg. when using x-pack)
+  # username = "telegraf"
+  # password = "mypassword"
+
+  ## Optional TLS Config
+  # tls_ca = "/etc/telegraf/ca.pem"
+  # tls_cert = "/etc/telegraf/cert.pem"
+  # tls_key = "/etc/telegraf/key.pem"
+  ## Use TLS but skip chain & host verification
+  # insecure_skip_verify = false
+
+[[inputs.elasticsearch_query.aggregation]]
+  measurement_name = "measurement"
+  index = "index-*"
+  date_field = "@timestamp"
+  ## Time window to query (eg. "1m" to query documents from last minute).
+  ## Normally should be set to same as collection interval
+  query_period = "1m"
+
+  ## Optional parameters:
+  ## Lucene query to filter results
+  filter_query = "*"
+  ## Fields to aggregate values (must be numeric fields)
+  metric_fields = ["metric"]
+  ## Aggregation function to use on the metric fields
+  ## Valid values are: avg, sum, min, max, sum
+  metric_function = "avg"
+  ## Fields to be used as tags
+  ## Must be non-analyzed fields, aggregations are performed per tag
+  tags = ["field.keyword", "field2.keyword"]
+  ## Set to true to not ignore documents when the tag(s) above are missing
+  include_missing_tag = true
+  ## String value of the tag when the tag does not exist
+  ## Used when include_missing_tag is true
+  missing_tag_value = "null"
+`
+
+type ElasticsearchQuery struct {
+	URLs                []string `toml:"urls"`
+	Username            string
+	Password            string
+	EnableSniffer       bool
+	Tracelog            bool
+	Timeout             internal.Duration
+	HealthCheckInterval internal.Duration
+	Aggregations        []Aggregation `toml:"aggregation"`
+	tls.ClientConfig
+	httpclient *http.Client
+	ESClient   *elastic.Client
+	acc        telegraf.Accumulator
```

Comment:
Usually the accumulator is passed down the functions and not stored in the structure. If you do not have a striking reason please follow this unwritten rule. :-)

Reply:
I think I did that way because of the `recurseResponse` function. I thought it was better to have it "globally" than passing down each recursion. I will try to revisit that.

Reply:
Ok that is a point, but maybe you can change the recursive function in a way that it returns what should be added to `acc` and not doing it while traversing the tree...